src/unity/python/turicreate/toolkits/object_detector/object_detector.py

```diff
@@ -265,11 +270,16 @@ def create(dataset, annotations=None, feature=None, model='darknet-yolo',
     anchors = params['anchors']
     num_anchors = len(anchors)
 
-    num_mxnet_gpus = _mxnet_utils.get_num_gpus_in_use(max_devices=params['batch_size'])
-    batch_size_each = params['batch_size'] // max(num_mxnet_gpus, 1)
+    if batch_size < 1:
+        batch_size = 32  # Default if not user-specified
+    num_mxnet_gpus = _mxnet_utils.get_num_gpus_in_use(max_devices=batch_size)
+    batch_size_each = batch_size // max(num_mxnet_gpus, 1)
     # Note, this may slightly alter the batch size to fit evenly on the GPUs
     batch_size = max(num_mxnet_gpus, 1) * batch_size_each
     use_mps = _use_mps() and num_mxnet_gpus == 0
+    if use_mps and _mps_device_memory_limit() < 4 * 1024 * 1024 * 1024:
+        # Reduce batch size for GPUs with less than 4GB RAM
+        batch_size = 16
```

Comment:
At face value, we may also need to reset `batch_size_each`, but AFAICT none of the references to this variable below are needed (in the `if use_mps` case).

What does the net.forward() call in line 422 do? Also, it looks like `ymap_shape` and `loss` are only used by the MXNet path. If so, moving those assignments to that path might make things a bit more clear

Reply:
Yes, the `batch_size_each` should only be relevant for multi-GPU. Although it's used to set `ymap_shape`, so for single-GPU we should just make that `batch_size_each == batch_size`. As you point out, I don't see a usage of `ymap_shape` in the MPS path right now, but it could be inviting future bugs to leave `batch_size` and `batch_size_each` inconsistent.

The `net.forward` should really have a comment explaining it. So, `net` here is an MXNet model with the right initial parameters that should be copied into MPS. However, the parameters may not have been initialized yet (since this is deferred until the first forward pass in MXNet). The call to `forward` forces this initialization.

Of course, all this is a bit of a hack. Ideally we should not be relying on MXNet at all to initialize the MPS network. However, getting to that point will take a bit of refactoring.

Reply:
Also, we should probably print the batch size if `verbose` is True, just like we show what `max_iterations` was automatically set to (as well as what computational resource was used).